File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/MiscUtils.java
package it.sephiroth.android.library.bottomnavigation;

import android.animation.Animator;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.ColorStateList;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.view.ViewCompat;
import android.support.v4.view.ViewPropertyAnimatorCompat;
import android.support.v4.view.ViewPropertyAnimatorListener;
import android.util.Log;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.animation.DecelerateInterpolator;

import it.sephiroth.android.library.bottonnavigation.R;

import static android.view.WindowManager.LayoutParams;

/**
 * Created by alessandro on 4/2/16.
 */
public final class MiscUtils {

    private MiscUtils() { }

    public static int getDimensionPixelSize(final Context context, final int dp) {
        return (int) (context.getResources().getDisplayMetrics().density * dp);
    }

    /**
     * Returns if the current theme has the translucent status bar enabled
     *
     * @param activity context
     * @return true if the current theme has the translucent statusbar
     */
    @TargetApi (19)
    public static boolean hasTranslucentStatusBar(@Nullable final Activity activity) {
        if (null == activity) {
            return false;
        }
        if (Build.VERSION.SDK_INT >= 19) {
            return
                ((activity.getWindow().getAttributes().flags & LayoutParams.FLAG_TRANSLUCENT_STATUS)
                    == LayoutParams.FLAG_TRANSLUCENT_STATUS);
        } else {
            return false;
        }
    }

    /**
     * Returns true if the current theme has declared the botton navigation as translucent
     *
     * @param activity context
     * @return true if the activity has the translucent navigation enabled
     */
    @TargetApi (19)
    public static boolean hasTranslucentNavigation(@Nullable final Activity activity) {
        if (null == activity) {
            return false;
        }
        if (Build.VERSION.SDK_INT >= 19) {
            return
                ((activity.getWindow().getAttributes().flags & LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
                    == LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
        } else {
            return false;
        }
    }

    /**
     * Returns the current theme defined color
     *
     * @param context
     * @param color
     * @return
     */
    protected static int getColor(Context context, @AttrRes int color) {
        TypedValue tv = new TypedValue();
        context.getTheme().resolveAttribute(color, tv, true);
        return tv.data;
    }

    @TargetApi (21)
    protected static void setDrawableColor(@NonNull final Drawable drawable, final int color) {
        if (Build.VERSION.SDK_INT >= 21) {
            if (RippleDrawable.class.isInstance(drawable)) {
                RippleDrawable rippleDrawable = (RippleDrawable) drawable;
                rippleDrawable.setColor(ColorStateList.valueOf(color));
            }
        } else {
            DrawableCompat.setTint(drawable, color);
        }
    }

    @SuppressLint ("RtlHardcoded")
    static boolean isGravitiyLeft(final int gravity) {
        return gravity == Gravity.LEFT;
    }

    @SuppressLint ("RtlHardcoded")
    static boolean isGravityRight(final int gravity) {
        return gravity == Gravity.RIGHT;
    }

    static boolean isGravityBottom(final int gravity) {
        return gravity == Gravity.BOTTOM;
    }

    protected static void switchColor(
        final BottomNavigation navigation,
        final View v,
        final View backgroundOverlay,
        final ColorDrawable backgroundDrawable,
        final int newColor) {

        backgroundOverlay.clearAnimation();

        if (Build.VERSION.SDK_INT >= 21) {
            Animator currentAnimator = (Animator) backgroundOverlay.getTag(R.id.bbn_backgroundOverlay_animator);
            if (null != currentAnimator) {
                currentAnimator.cancel();
            }
        }

        backgroundDrawable.setColor(newColor);
        backgroundOverlay.setVisibility(View.INVISIBLE);
        ViewCompat.setAlpha(backgroundOverlay, 1);
    }

    protected static void animate(
        final BottomNavigation navigation, final View v, final View backgroundOverlay, final ColorDrawable backgroundDrawable,
        final int newColor, final long duration) {

        int centerX = (int) (ViewCompat.getX(v) + (v.getWidth() / 2));
        int centerY = navigation.getPaddingTop() + v.getHeight() / 2;

        backgroundOverlay.clearAnimation();

        final Object animator;

        if (Build.VERSION.SDK_INT >= 21) {

            Animator currentAnimator = (Animator) backgroundOverlay.getTag(R.id.bbn_backgroundOverlay_animator);
            if (null != currentAnimator) {
                currentAnimator.cancel();
            }

            final float startRadius = 10;
            final float finalRadius = centerX > navigation.getWidth() / 2 ? centerX : navigation.getWidth() - centerX;
            animator =
                ViewAnimationUtils.createCircularReveal(backgroundOverlay, centerX, centerY, startRadius, finalRadius);
            backgroundOverlay.setTag(R.id.bbn_backgroundOverlay_animator, animator);
        } else {
            ViewCompat.setAlpha(backgroundOverlay, 0);
            animator = ViewCompat.animate(backgroundOverlay).alpha(1);
        }

        backgroundOverlay.setBackgroundColor(newColor);
        backgroundOverlay.setVisibility(View.VISIBLE);

        if (ViewPropertyAnimatorCompat.class.isInstance(animator)) {
            ((ViewPropertyAnimatorCompat) animator).setListener(new ViewPropertyAnimatorListener() {
                boolean cancelled;

                @Override
                public void onAnimationStart(final View view) { }

                @Override
                public void onAnimationEnd(final View view) {
                    if (!cancelled) {
                        backgroundDrawable.setColor(newColor);
                        backgroundOverlay.setVisibility(View.INVISIBLE);
                        ViewCompat.setAlpha(backgroundOverlay, 1);
                    }
                }

                @Override
                public void onAnimationCancel(final View view) {
                    cancelled = true;
                }
            })
                .setDuration(duration)
                .start();
        } else {
            Animator animator1 = (Animator) animator;
            animator1.setDuration(duration);
            animator1.setInterpolator(new DecelerateInterpolator());
            animator1.addListener(new Animator.AnimatorListener() {
                boolean cancelled;

                @Override
                public void onAnimationStart(final Animator animation) { }

                @Override
                public void onAnimationEnd(final Animator animation) {
                    if (!cancelled) {
                        backgroundDrawable.setColor(newColor);
                        backgroundOverlay.setVisibility(View.INVISIBLE);
                        ViewCompat.setAlpha(backgroundOverlay, 1);
                    }
                }

                @Override
                public void onAnimationCancel(final Animator animation) {
                    cancelled = true;
                }

                @Override
                public void onAnimationRepeat(final Animator animation) { }
            });

            animator1.start();
        }
    }

    public static void log(final String tag, final int level, String message, Object... arguments) {
        if (BottomNavigation.DEBUG) {
            Log.println(level, tag, String.format(message, arguments));
        }
    }

    @Nullable
    static Activity getActivity(@Nullable Context context) {
        if (context == null) {
            return null;
        } else if (context instanceof Activity) {
            return (Activity) context;
        } else if (context instanceof ContextWrapper) {
            return getActivity(((ContextWrapper) context).getBaseContext());
        }
        return null;
    }

}

File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/MiscUtils.kt
package it.sephiroth.android.library.bottomnavigation

import android.animation.Animator
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.content.res.ColorStateList
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.os.Build
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewAnimationUtils
import android.view.WindowManager.LayoutParams
import android.view.animation.DecelerateInterpolator
import androidx.annotation.AttrRes
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.view.ViewCompat
import androidx.core.view.ViewPropertyAnimatorCompat
import androidx.core.view.ViewPropertyAnimatorListener
import it.sephiroth.android.library.bottonnavigation.R
import timber.log.Timber

/**
 * Created by alessandro on 4/2/16.
 */
@Suppress("unused", "UNUSED_PARAMETER")
object MiscUtils {

    fun getDimensionPixelSize(context: Context, dp: Int): Int {
        return (context.resources.displayMetrics.density * dp).toInt()
    }

    /**
     * Returns if the current theme has the translucent status bar enabled
     *
     * @param activity context
     * @return true if the current theme has the translucent statusbar
     */
    @TargetApi(19)
    fun hasTranslucentStatusBar(activity: Activity?): Boolean {
        if (null == activity) {
            return false
        }
        return if (Build.VERSION.SDK_INT >= 19) {
            activity.window.attributes.flags and LayoutParams.FLAG_TRANSLUCENT_STATUS == LayoutParams.FLAG_TRANSLUCENT_STATUS
        } else {
            false
        }
    }

    /**
     * Returns true if the current theme has declared the botton navigation as translucent
     *
     * @param activity context
     * @return true if the activity has the translucent navigation enabled
     */
    @TargetApi(19)
    fun hasTranslucentNavigation(activity: Activity?): Boolean {
        if (null == activity) {
            return false
        }
        return if (Build.VERSION.SDK_INT >= 19) {
            activity.window.attributes.flags and LayoutParams.FLAG_TRANSLUCENT_NAVIGATION == LayoutParams.FLAG_TRANSLUCENT_NAVIGATION
        } else {
            false
        }
    }

    /**
     * Returns the current theme defined color
     *
     * @param context
     * @param color
     * @return
     */
    fun getColor(context: Context, @AttrRes color: Int): Int {
        val tv = TypedValue()
        context.theme.resolveAttribute(color, tv, true)
        return tv.data
    }

    @TargetApi(21)
    fun setDrawableColor(drawable: Drawable, color: Int) {
        if (Build.VERSION.SDK_INT >= 21 && drawable is RippleDrawable) {
            if (RippleDrawable::class.java.isInstance(drawable)) {
                drawable.setColor(ColorStateList.valueOf(color))
            }
        } else {
            DrawableCompat.setTint(drawable, color)
        }
    }

    @SuppressLint("RtlHardcoded")
    fun isGravitiyLeft(gravity: Int): Boolean {
        return gravity == Gravity.LEFT
    }

    @SuppressLint("RtlHardcoded")
    fun isGravityRight(gravity: Int): Boolean {
        return gravity == Gravity.RIGHT
    }

    fun isGravityBottom(gravity: Int): Boolean {
        return gravity == Gravity.BOTTOM
    }

    fun switchColor(
            navigation: BottomNavigation,
            v: View,
            backgroundOverlay: View,
            backgroundDrawable: ColorDrawable,
            newColor: Int) {

        backgroundOverlay.clearAnimation()

        if (Build.VERSION.SDK_INT >= 21) {
            val currentAnimator = backgroundOverlay.getTag(R.id.bbn_backgroundOverlay_animator) as Animator?
            currentAnimator?.cancel()
        }

        backgroundDrawable.color = newColor
        backgroundOverlay.visibility = View.INVISIBLE
        @Suppress("DEPRECATION")
        ViewCompat.setAlpha(backgroundOverlay, 1f)
    }

    fun animate(
            navigation: BottomNavigation, v: View, backgroundOverlay: View, backgroundDrawable: ColorDrawable,
            newColor: Int, duration: Long) {

        val centerX = (v.x + v.width / 2).toInt()
        val centerY = navigation.paddingTop + v.height / 2

        backgroundOverlay.clearAnimation()

        val animator: Any

        if (Build.VERSION.SDK_INT >= 21) {

            val currentAnimator = backgroundOverlay.getTag(R.id.bbn_backgroundOverlay_animator) as Animator?
            currentAnimator?.cancel()

            val startRadius = 10f
            val finalRadius = (if (centerX > navigation.width / 2) centerX else navigation.width - centerX).toFloat()
            animator = ViewAnimationUtils.createCircularReveal(backgroundOverlay, centerX, centerY, startRadius, finalRadius)
            backgroundOverlay.setTag(R.id.bbn_backgroundOverlay_animator, animator)
        } else {
            @Suppress("DEPRECATION")
            ViewCompat.setAlpha(backgroundOverlay, 0f)
            animator = ViewCompat.animate(backgroundOverlay).alpha(1f)
        }

        backgroundOverlay.setBackgroundColor(newColor)
        backgroundOverlay.visibility = View.VISIBLE

        if (ViewPropertyAnimatorCompat::class.java.isInstance(animator)) {
            (animator as ViewPropertyAnimatorCompat).setListener(object : ViewPropertyAnimatorListener {
                var cancelled: Boolean = false

                override fun onAnimationStart(view: View) {}

                override fun onAnimationEnd(view: View) {
                    if (!cancelled) {
                        backgroundDrawable.color = newColor
                        backgroundOverlay.visibility = View.INVISIBLE
                        @Suppress("DEPRECATION")
                        ViewCompat.setAlpha(backgroundOverlay, 1f)
                    }
                }

                override fun onAnimationCancel(view: View) {
                    cancelled = true
                }
            })
                .setDuration(duration)
                .start()
        } else {
            val animator1 = animator as Animator
            animator1.duration = duration
            animator1.interpolator = DecelerateInterpolator()
            animator1.addListener(object : Animator.AnimatorListener {
                var cancelled: Boolean = false

                override fun onAnimationStart(animation: Animator) {}

                override fun onAnimationEnd(animation: Animator) {
                    if (!cancelled) {
                        backgroundDrawable.color = newColor
                        backgroundOverlay.visibility = View.INVISIBLE
                        @Suppress("DEPRECATION")
                        ViewCompat.setAlpha(backgroundOverlay, 1f)
                    }
                }

                override fun onAnimationCancel(animation: Animator) {
                    cancelled = true
                }

                override fun onAnimationRepeat(animation: Animator) {}
            })

            animator1.start()
        }
    }

    fun log(level: Int, message: String, vararg arguments: Any) {
        if (BottomNavigation.DEBUG) {
            Timber.log(level, message, arguments)
        }
    }

    fun getActivity(context: Context?): Activity? {
        return when (context) {
            null -> null
            is Activity -> context
            is ContextWrapper -> getActivity(context.baseContext)
            else -> null
        }
    }

}
